### PR TITLE
Retry customization

### PR DESCRIPTION
**Pull request by ivanprado**
In this PR:
* Customization of headers in `request_raw`
* Customization of retry policy in `request_raw`
* Improved stats (counting billable responses, more meaningful success ratio)
* Fix a bug with retry messages (some of them were missing)

### CODE REVIEW

File: autoextract/aio/client.py [hunk 1]
```diff
@@ -9,24 +9,26 @@
...
                                         sock_read=API_TIMEOUT + 30,
                                         sock_connect=10)
 
 
-def create_session(**kwargs) -> aiohttp.ClientSession:
+def create_session(connection_pool_size=100, **kwargs) -> aiohttp.ClientSession:
     """ Create a session with parameters suited for AutoExtract """
     kwargs.setdefault('timeout', AIO_API_TIMEOUT)
+    if "connector" not in kwargs:
+        kwargs["connector"] = TCPConnector(limit=connection_pool_size)
```
Comment (kmike) [inline autoextract/aio/client.py:31]: Why is this needed? I was hoping that we're using http 1.1 and sending all requests in a single tcp connection; we're connectingto a single host.

Comment (ivanprado) [reply autoextract/aio/client.py:31]: @kmike This is a good point. I can say that I have tested to set TCPConnector(limit=1) and performance is dramatically affected. It seems requests are sent in sequence in such a case. Are you sure that under Http1.1 it is possible to mix request/responses under the same connection in a parallel way? I mean, requests are never blocked, and responses are being read as they come (some kind of streaming). From [Wikipedia](https://en.wikipedia.org/wiki/Chunked_transfer_encoding):

> Chunked encoding has the benefit that it is not necessary to generate the full content before writing the header, as it allows streaming of content as chunks and explicitly signaling the end of the content, **making the connection available for the next HTTP request/response**.

According to that still under Http1.1 requests are executed in sequence it seems. So this is why the pool size matters.

Comment (kmike) [reply autoextract/aio/client.py:31]: I think you're right @ivanprado, and that's a correct change.

File: autoextract/aio/client.py [hunk 2]
```diff
@@ -175,10 +202,17 @@ async def request_raw(query: Query,
     ``agg_stats`` argument allows to keep track of various stats; pass an
     ``AggStats`` instance, and it'll be updated.
 
+    Additional ``headers`` for the requests can be provided.
+
+    Default retry policy can be customized by providing a
+    ``retry_wrapper`` that can be built with a customized
+    :class:`autoextract.retry.RetryFactory`
```
Comment (kmike) [inline autoextract/aio/client.py:209]: Why is it called retry_wrapper? As I understand, this function receives arbitrary tenacity.Retrying object; autoextract.retry.RetryFactory can be used as a starting point/helper to build customized Retrying object for AutoExtract.

File: autoextract/aio/retry.py [hunk 3]
```diff
@@ -64,92 +63,105 @@ def _is_retriable_query_error(exc: Exception) -> bool:
...
-    def __call__(self, retry_state: RetryCallState) -> float:
+class RetryFactory:
+    """
+    Build custom retry configuration
+    """
+    retry_condition = (
+            retry_if_exception(_is_throttling_error) |
+            retry_if_exception(_is_network_error) |
+            retry_if_exception(_is_server_error) |
+            retry_if_exception(_is_retriable_query_error)
```
Comment (kmike) [inline autoextract/aio/retry.py:82]:
```suggestion
        retry_if_exception(_is_throttling_error) |
        retry_if_exception(_is_network_error) |
        retry_if_exception(_is_server_error) |
        retry_if_exception(_is_retriable_query_error)
```

File: autoextract/aio/retry.py [hunk 4]
```diff
@@ -64,92 +63,105 @@ def _is_retriable_query_error(exc: Exception) -> bool:
...
+    retryable_query_error_stop = stop_after_delay(15 * 60)
+
+    def __init__(self, **kwargs):
+        mutable_attributes = {attrib for attrib in dir(self)
+                              if not attrib.startswith("_")}
+        for k, v in kwargs:
+            if not k in mutable_attributes:
+                raise TypeError(f"Attribute {k} is invalid. Any of "
+                                f"{mutable_attributes} was expected")
+            setattr(self, k, v)
```
Comment (kmike) [inline autoextract/aio/retry.py:116]: Is this a shortcut to allow setting some of the attributes from the constructor?
If so, it'd be good to have tests for it.
However, maybe we can remove this feature? One can subclass RetryFactory and change an attribute this way, should be easy enough, and won't require code in `__init__`.

**Comment by kmike** [comment]
`tests/__init__.py` file was omitted on purpose, see https://docs.pytest.org/en/stable/goodpractices.html#tests-outside-application-code

File: autoextract/aio/client.py [hunk 5]
```diff
@@ -175,10 +202,17 @@ async def request_raw(query: Query,
     ``agg_stats`` argument allows to keep track of various stats; pass an
     ``AggStats`` instance, and it'll be updated.
 
+    Additional ``headers`` for the requests can be provided.
```
Comment (kmike) [inline autoextract/aio/client.py:205]: I think it makes sense to clarify that these are headers sent to the API server, not to remote websites.

File: autoextract/aio/retry.py (hunk 4 above)
Comment (ivanprado) [reply autoextract/aio/retry.py]: Sure, sure! I thought I have already removed it. Agree with you, no `__init__` at all is required here. Removed here: https://github.com/scrapinghub/scrapinghub-autoextract/pull/24/commits/62dea1775a3afe8ca136f7e437cebe649b74e689

File: autoextract/aio/client.py (hunk 2 above)
Comment (kmike) [reply autoextract/aio/client.py:209]: What do you think about calling it `retry_policy`, `retry_strategy` or something like that? Document that it should be a tenacity.Retrying instance, and that :class:`autoextract.retry.RetryFactory` helper exist.

Comment (ivanprado) [reply autoextract/aio/client.py:209]: Yeah, I was in doubt about how to name it. The reason to call it `retry_wrapper` is that it is actually just a function wrapper what is required. It doesn't even require to be a tenacity wrapper. See the tenacity code here: https://github.com/jd/tenacity/blob/master/tenacity/__init__.py#L109 

> Wrap a function with a new `Retrying` object.

A wrapper like the following would also be acceptable:
```py
def simple_retry(count):
  def fn(f):
    for _ in range(count):
      try:
        return f()
      except Exception as e:
        pass
    raise e
```
So you could pass `simple_rety` as the `retry_wrapper`.

If you have a suggestion about a better name it will be welcome!

File: autoextract/aio/client.py (hunk 5 above)
Comment (ivanprado) [reply autoextract/aio/client.py]: Make sense. Done here: https://github.com/scrapinghub/scrapinghub-autoextract/pull/24/commits/94484d498961c016d201966366b4076b13f009c9

**Comment by ivanprado** [comment]
> tests/__init__.py file was omitted on purpose, see https://docs.pytest.org/en/stable/goodpractices.html#tests-outside-application-code

@kmike I had no idea about that :+1: . The file was added automatically by PyCharm and I left it there. Removed in https://github.com/scrapinghub/scrapinghub-autoextract/pull/24/commits/bb8c8d858c65e0087d567de70b1a48b92f3266a8. I have read the link, but I didn't get the idea why is this a bad practice. Anyway, now I know I should not do that :-)

File: autoextract/aio/client.py (hunk 5 above)
Comment (ivanprado) [reply autoextract/aio/client.py]: Done here: https://github.com/scrapinghub/scrapinghub-autoextract/pull/24/commits/94484d498961c016d201966366b4076b13f009c9

File: autoextract/aio/client.py (hunk 2 above)
Comment (ivanprado) [reply autoextract/aio/client.py:212]: Ohh, sorry, now I see your suggestions: `retry_policy`, `retry_strategy`. I'll think about it.

Comment (ivanprado) [reply autoextract/aio/client.py]: Ok, so finally I decided to clarify the documentation:

https://github.com/scrapinghub/scrapinghub-autoextract/pull/24/commits/ebcb1318687d28a612108fae0831b3ba3e00b5f2

**Comment by ivanprado** [comment]
@kmike I went over all the comments and I'm ready for a re-review.

File: autoextract/aio/client.py [hunk 6]
```diff
@@ -175,10 +202,41 @@ async def request_raw(query: Query,
...
+
+        def retry_3_attempts(request_fn):
+            async def fn(*args, **kwargs):
+                for _ in range(3):
+                    try:
+                        return await request_fn()
+                    except Exception as e:
+                        exc = e
+                raise exc
+            return fn
```
Comment (kmike) [inline autoextract/aio/client.py:224]: The docs make it much more clear, but I'm still unsure about the API.
So, we're saying in this example:

1. retry_wrapper wraps some unspecified internal request function
2. Signature of this internal function is not documented. 
3. From the example you can figure out that this internal function has no arguments, must be awaited (i.e. it is a coroutine), and returns something which the wrapper should return. Also, in the example wrapper receives *args and **kwargs for some reason.

I think it is a bit under-specified. But at the same time, I'm not sure what's an advatnage of specifying it further - documenting internal function signature, etc. 

Won't saying that we receive tenacity.Retrying object cover the same use cases, require less documentation, and be more clear for users? Or do you see use cases for passing some other request wrappers here?

Comment (ivanprado) [reply autoextract/aio/client.py:224]: I see your point. Ideally, we should provide even a much simpler retry configuration interface. E.g. a `RetryConfig` object with some simple possibilities, like `number_of_attemps`, `max_time`, `wait_range`. Something very simple to configure and flexible enough. But this was not my goal with this change. My goal here with retrials was been able to provide maximum flexibility for very advances usages (e.g. being able to change logging by intercepting, etc). So I know in advance that this is a bit obscure. I'll change it to `tenacity.AsyncRetrying` as you suggest, it might improve a bit but it will still be obscure and under-documented, not for the "regular" user. I'm going to work on it. 

> Or do you see use cases for passing some other request wrappers here?

In reality, no. 

> Won't saying that we receive tenacity.Retrying object cover the same use cases, require less documentation, and be more clear for users?

Yes, it will probably improve. Let me change it.

Comment (ivanprado) [reply autoextract/aio/client.py]: @kmike done here: https://github.com/scrapinghub/scrapinghub-autoextract/pull/24/commits/136af5acafc821cb862c1e71dd76e7dcf8d65472 . It seems rounder :+1:

File: autoextract/aio/retry.py [hunk 7]
```diff
@@ -155,4 +155,4 @@ def build(self):
         )
 
 
-autoextract_retry = RetryFactory().build()
+autoextract_retrying: AsyncRetrying = RetryFactory().build()
```
Comment (kmike) [inline autoextract/aio/retry.py:158]: Probably the last question/nitpick :) I *think* that so far there was nothing about async in this retry.py module. The idea was to reuse it in a sync part as well; it is a nice thing about tenacity that it allows to decouple the strategy of retrying from the way it is applied.

Comment (kmike) [reply autoextract/aio/retry.py:158]: If I'm not mistaken, autoextract_retrying is a wrapper, not a AsyncRetrying instance. I was wrong about Retrying / AsyncRetrying thing, we're not using them directly.

**Comment by ivanprado** [comment]
@kmike I have included this piece of code https://github.com/scrapinghub/scrapinghub-autoextract/pull/24/commits/e2b96233e243070ab1ee3c7496ee43733a4db3e5 in the PR, as it will be required in `autoextract_providers` but here is the right place for it.

File: autoextract/aio/retry.py (hunk 7 above)
Comment (ivanprado) [reply autoextract/aio/retry.py:158]: This is the place where tenacity decides which `Retrying` to use: https://github.com/jd/tenacity/blob/master/tenacity/__init__.py#L120-L126 . The result is a wrapper. 

Before the last commit, `autoextract_retry` was a wrapper created with `retry`, and then compatible with sync/async. But the last commit breaks the compatibility: explicit `AsyncRetrying` was introduced. I don't see an elegant way of keeping both things (explicit `Retrying` but with autodetection of sync/async). 

@kmike So then should I revert the commit and leave it as it was before?

Comment (kmike) [reply autoextract/aio/retry.py:158]: I see, thanks for the explanation. I think that's ffine to keep it as-is. If we decide to reuse retries for the sync part, we can have build_sync / build_async methods.

**Comment by kmike** [comment]
Thanks @ivanprado!

**Comment by ivanprado** [comment]
Thank you @kmike for the review!

File: autoextract/stats.py [hunk 8]
```diff
@@ -49,7 +76,7 @@ def error_ratio(self):
 
     @zero_on_division_error
     def success_ratio(self):
-        return self.n_results / (self.n_results + self.n_fatal_errors)
+        return self.n_extracted_queries / self.n_input_queries
```
Comment (kmike) [inline autoextract/stats.py:79]: Hey @ivanprado! I tried using the CLI with the updated code, and this changes the meaning of success in the progress bar. I wanted success to be defined as "% of input queries which we were able to get the results for" - basically, success rate as measured in the output file. It seems now this % is lower, as it includes unsuccessful retries. 

For example, if there are 100 input queries, each required 2 attempts, but it was possible to get everything, previously success was 100% (as we were able to get all results). Now, if I'm not mistaken, success is going to be 50% (as half of requests were retries).

On the other hand, it'd be nice to include query errors to success calculation. Previously they were not included.

So, proposal:

1. Define success as % of non-error results in the output file.
2. If that's important, measure "retry overhead" separately.

What do you think?

Comment (kmike) [reply autoextract/stats.py:79]: Another example: user increases max-queery-retries. This is likely to increase % of extracted results, but also increase % of retries even more. So a lower success rate will be reported in the CLI, which may motivate user not to enable retries.

Comment (ivanprado) [reply autoextract/stats.py:79]: > I wanted success to be defined as "% of input queries which we were able to get the results for"

Agree. This is what I wanted the code to do, but maybe there is a bug anywhere. 

> It seems now this % is lower, as it includes unsuccessful retries.

Why does it include unsuccessful retries? `n_input_queries` should be only counting input queries and skipping the retries logic here: https://github.com/scrapinghub/scrapinghub-autoextract/blob/master/autoextract/aio/client.py#L250 

Besides `n_extracted_queries` should only be counting successful responses, which are not being retried. So, I don't see retries affecting here as well. 

Regarding your proposal, point number (1) I think it is already achieved (unless there is a bug I'm not able to see). And point number (2) of "retry overhead" is something is also being computed. See the following example stats:

 ```
Summary
-------
Mean connection time:     0.35
Mean response time:       10.07
Throttle ratio:           0.0%
Attempts:                 11
Errors:                   90.9%, fatal: 0, non fatal: 10
Successful URLs:          2 of 11
Success ratio:            18.2%
Billable query responses: 18 of 19
```

That's the way of reading them: Sucess ratio is 18% because only 2 of the 11 URLs in the input file were successfully extracted (free of any kind of error). But it required sending 19 queries in 11 request attempts (batch size is 2 here). So the overhead in queries here is 19/11, and the overhead in requests is 11/6 (6 requests it the minimum number of requests possible if no errors to obtain 11 queries responses).

Something that can be confusing is that `n_input_queries` computing is ahead of `n_extracted_queries` so **the impression, if you don't wait until the end, is that success rate is way low**. But it will eventually catch up at the end. I didn't find a quick solution for that so I left as is, but maybe there is a way to improve it somehow. 

Finally, I think there is room for improvement in the the `tqdm` error section. I don't understand the distinction between `erros` and `fatal_errors`, and also it will be very confusing for the user. But I didn't change things there as I didn't get the full idea of the errors (and I full of work for this week, I decided not going deeper ;-) ).

Comment (ivanprado) [reply autoextract/stats.py:79]: I fixed the problem of showing underestimated success rate in `tqdm` in this PR: https://github.com/scrapinghub/scrapinghub-autoextract/pull/25/files. It was just a matter of moving a line of code. Thank you @kmike for the inspiration.
